GitCommands/Settings/Setting.cs

```diff
@@ -4,91 +4,129 @@
 
 namespace GitCommands.Settings
 {
-    public abstract class Setting<T>
+    public interface ISetting<T>
     {
-        public readonly SettingsPath SettingsSource;
-        public readonly T DefaultValue;
-        public readonly string Name;
+        event EventHandler Updated;
 
-        public event EventHandler Updated;
+        SettingsPath SettingsSource { get; }
 
-        protected Setting(string name, SettingsPath settingsSource, T defaultValue)
+        string Name { get; }
+
+        T DefaultValue { get; }
+
+        T ValueOrDefault { get; }
+
+        T Value { get; set; }
+
+        string FullPath { get; }
+    }
+
+    public static class Setting
+    {
+        public static ISetting<string> Create(SettingsPath settingsSource, string name, string defaultValue)
         {
-            Name = name;
-            SettingsSource = settingsSource;
-            DefaultValue = defaultValue;
+            return new SettingOf<string>(settingsSource, name, defaultValue);
         }
 
-        public virtual T Value
+        public static ISetting<T> Create<T>(SettingsPath settingsSource, string name, T defaultValue)
+            where T : struct
         {
-            get => GetValue(Name, DefaultValue);
-            set
-            {
-                if (Value.Equals(value))
-                {
-                    return;
-                }
+            return new SettingOf<T>(settingsSource, name, defaultValue);
+        }
 
-                SetValue(Name, value);
-                Updated?.Invoke(this, EventArgs.Empty);
-            }
+        public static ISetting<T?> Create<T>(SettingsPath settingsSource, string name)
+            where T : struct
+        {
+            return new SettingOf<T?>(settingsSource, name);
         }
 
-        public T ValueOrDefault
+        private sealed class SettingOf<T> : ISetting<T>
         {
-            get
+            public event EventHandler Updated;
+
+            public SettingOf(SettingsPath settingsSource, string name, T defaultValue = default)
             {
-                T v = Value;
-                if (ValueIsEmpty(v))
-                {
-                    return DefaultValue;
-                }
-                else
-                {
-                    return v;
-                }
+                SettingsSource = settingsSource;
+                Name = name;
+                DefaultValue = defaultValue;
             }
-        }
 
-        public virtual bool ValueIsEmpty(T value)
-        {
-            return EqualityComparer<T>.Default.Equals(value, default);
-        }
+            public SettingsPath SettingsSource { get; }
 
-        public string FullPath => SettingsSource.PathFor(Name);
+            public string Name { get; }
 
-        private T GetValue(string name, T defaultValue = default)
-        {
-            return SettingsSource.GetValue(name, defaultValue, value =>
+            public T DefaultValue { get; }
```

Comment:
This name a bugs me a little.
There is: `Value` and `ValueOrDefault`, so it feels the 3rd options should be just called `Default`. Especially since we have `IsDefault` too.

Reply:
Controversial purpose of this property.

For non nullable Value == DefaultValue
For nullable Value == DefaultValue, cause null is value too.

Reply:
Agree, that's why typically nullable and non-nullable API have similar yet different implementations.
In this case I wouldn't spend too much time on it though - apply the comment, rename and move on.